op-batcher/batcher/espresso.go

```diff
@@ -39,140 +35,285 @@ const (
+		return fmt.Errorf("failed to create Espresso transaction from a batch: %w", err)
 	}
 
-	namespaceOk := espressoVerification.VerifyNamespace(
-		exampleNamespace,
-		resp.Proof,
-		*header.Header.GetPayloadCommitment(),
-		*header.Header.GetNsTable(),
-		resp.Transactions,
-		resp.VidCommon,
-	)
-
-	if !namespaceOk {
-		return fmt.Errorf("error validating namespace proof (height: %d)", height)
-	}
+	go func() {
+		// We will retry publishing until successful
+		for {
+			err := l.tryPublishBatchToEspresso(ctx, *transaction)
+			if err == nil {
+				l.Log.Info(fmt.Sprintf("Published block %s to Espresso", eth.ToBlockID(block)))
+				break
+			}
+		}
+	}()
 
 	return nil
 }
 
-func (l *BatchSubmitter) submitToEspresso(txdata txData, sig, batcherSignature []byte) (*EspressoCommitment, error) {
-	transaction := Transaction{
-		Namespace:        exampleNamespace,
-		BatcherSignature: batcherSignature,
-		CallData:         txdata.CallData(),
-	}.toEspresso()
-	txHash, err := l.Espresso.SubmitTransaction(l.shutdownCtx, transaction)
-	if err != nil {
-		l.Log.Error("Failed to submit transaction", "transaction", transaction, "error", err)
-		l.recordFailedDARequest(txdata.ID(), err)
-		return nil, fmt.Errorf("failed to submit transaction: %w", err)
+func (l *BatchSubmitter) espressoSyncAndRefresh(ctx context.Context, newSyncStatus *eth.SyncStatus, streamer *espresso.EspressoStreamer) {
+	shouldClearState, err := streamer.Refresh(ctx, newSyncStatus)
+	shouldClearState = shouldClearState || err != nil
+
+	l.channelMgrMutex.Lock()
+	defer l.channelMgrMutex.Unlock()
+	syncActions, outOfSync := computeSyncActions(*newSyncStatus, l.prevCurrentL1, l.channelMgr.blocks, l.channelMgr.channelQueue, l.Log, l.Config.PreferLocalSafeL2)
+	if outOfSync {
+		l.Log.Warn("Sequencer is out of sync, retrying next tick.")
+		return
 	}
+	l.prevCurrentL1 = newSyncStatus.CurrentL1
+	if syncActions.clearState != nil || shouldClearState {
+		l.channelMgr.Clear(*syncActions.clearState)
+	} else {
+		l.channelMgr.PruneSafeBlocks(syncActions.blocksToPrune)
+		l.channelMgr.PruneChannels(syncActions.channelsToPrune)
+	}
+}
 
-	timer := time.NewTimer(transactionFetchTimeout)
-	defer timer.Stop()
+// Periodically refreshes the sync status and polls Espresso streamer for new batches
+func (l *BatchSubmitter) espressoBatchLoadingLoop(ctx context.Context, wg *sync.WaitGroup, publishSignal chan struct{}) {
+	l.Log.Info("Starting EspressoBatchLoadingLoop")
 
-	ticker := time.NewTicker(transactionFetchInterval)
+	defer wg.Done()
+	ticker := time.NewTicker(l.Config.PollInterval)
 	defer ticker.Stop()
 
-	var txQueryData espressoCommon.TransactionQueryData
-Loop:
+	streamer := espresso.EspressoStreamer{
+		BatcherAddress: l.BatcherAddress,
+		Namespace:      l.RollupConfig.L2ChainID.Uint64(),
+
+		L1Client:            l.L1Client,
+		EspressoClient:      l.Espresso,
+		EspressoLightClient: l.EspressoLightClient,
+		Log:                 l.Log,
+
+		BatchPos: 1,
```

Comment:
Why 1?

Reply:
Because it's the block number of the batch we want to give out next